retz-server/src/main/java/io/github/retz/web/JobRequestHandler.java

```diff
@@ -172,22 +173,24 @@ static String downloadFile(spark.Request req, spark.Response res) throws IOExcep
         LOG.debug("download: path={}", file);
 
         if (job.isPresent() && job.get().url() != null) { // If url() is null, the job hasn't yet been started at Mesos
-
-            // TODO: This is soo inefficient; it stores all data into RAM without streaming
-            // THIS MUST BE FIXED SOON
-            Pair<Integer, byte[]> payload = MesosHTTPFetcher.downloadHTTPFile(job.get().url(), file);
-            res.status(payload.left());
-
-            if (payload.left() == 200) {
-                res.type("application/octet-stream");
-                res.raw().setContentLength(payload.right().length);
-                ByteArrayInputStream in = new ByteArrayInputStream(payload.right());
-
-                IOUtils.copy(in, res.raw().getOutputStream());
-                in.close();
-            }
+            MesosHTTPFetcher.downloadHTTPFile(job.get().url(), file, (Triad<Integer, String, Pair<Long, InputStream>> triad) -> {
+                Integer statusCode = triad.left();
+                res.status(statusCode);
+                if (statusCode == 200) {
+                    Long length = triad.right().left();
+                    InputStream io = triad.right().right();
+                    Long maxFileSize = scheduler.get().maxFileSize();
+                    if (0 < maxFileSize && maxFileSize < length) {
+                        throw new IOException("Stream file too large: " + length + " < " + maxFileSize);
+                    }
+                    res.raw().setHeader("Content-Length", length.toString());
+                    LOG.info("start streaming of {} bytes", length);
+                    IOUtils.copyLarge(io, res.raw().getOutputStream());
```

Comment:
> In case of large files, I'd prefer public static long copyLarge(Reader input, Writer output, long inputOffset, long length) to prevent endless streaming, although Mesos agents could rarely be buggy or malicious.

Do you mean we should specify `inputOffset` and `length` or use `copyLarge(Reader input, Writer output, ...)` instead of `copyLarge(InputStream input, OutputStream output, ...)`?